mobile/lib/shared/providers/asset.provider.dart

```diff
@@ -220,80 +120,79 @@ class AssetNotifier extends StateNotifier<AssetsState> {
     }
-
-    return newAsset.isFavorite;
   }
 
-  Future<void> toggleArchive(Iterable<Asset> assets, bool status) async {
-    final newAssets = await Future.wait(
-      assets.map((a) => _assetService.changeArchiveStatus(a, status)),
-    );
+  Future<void> toggleArchive(List<Asset> assets, bool status) async {
+    final newAssets = await _assetService.changeArchiveStatus(assets, status);
     int i = 0;
-    bool unArchived = false;
     for (Asset oldAsset in assets) {
       final newAsset = newAssets[i++];
       if (newAsset == null) {
         log.severe("Change archive status failed for asset ${oldAsset.id}");
         continue;
       }
-      final index = state.allAssets.indexWhere((a) => oldAsset.id == a.id);
-      if (newAsset.isArchived) {
-        // remove from state
-        if (index != -1) {
-          state.allAssets.removeAt(index);
-        }
-      } else {
-        // add to state is difficult because the list is sorted
-        unArchived = true;
-      }
-    }
-    if (unArchived) {
-      final User me = Store.get(StoreKey.currentUser);
-      await _stateUpdateLock.run(
-        () async => _updateAssetsState(await _getUserAssets(me.isarId)),
-      );
-    } else {
-      _updateAssetsState(state.allAssets);
     }
   }
 }
 
 final assetProvider = StateNotifierProvider<AssetNotifier, AssetsState>((ref) {
   return AssetNotifier(
     ref.watch(assetServiceProvider),
-    ref.watch(appSettingsServiceProvider),
     ref.watch(albumServiceProvider),
     ref.watch(syncServiceProvider),
     ref.watch(dbProvider),
   );
 });
 
-final assetGroupByMonthYearProvider = StateProvider((ref) {
-  // TODO: remove `where` once temporary workaround is no longer needed (to only
-  // allow remote assets to be added to album). Keep `toList()` as to NOT sort
-  // the original list/state
-  final assets =
-      ref.watch(assetProvider).allAssets.where((e) => e.isRemote).toList();
+final assetDetailProvider =
+    StreamProvider.autoDispose.family<Asset, Asset>((ref, asset) async* {
+  yield await ref.watch(assetServiceProvider).loadExif(asset);
+  final db = ref.watch(dbProvider);
+  await for (final a in db.assets.watchObject(asset.id)) {
+    if (a != null) yield await ref.watch(assetServiceProvider).loadExif(a);
+  }
+});
 
-  assets.sortByCompare<DateTime>(
-    (e) => e.fileCreatedAt,
-    (a, b) => b.compareTo(a),
-  );
+final assetsProvider = StreamProvider.autoDispose<RenderList>((ref) async* {
+  final query = ref
+      .watch(dbProvider)
+      .assets
+      .filter()
+      .ownerIdEqualTo(Store.get(StoreKey.currentUser).isarId)
+      .isArchivedEqualTo(false)
+      .sortByFileCreatedAtDesc();
+  final settings = ref.watch(appSettingsServiceProvider);
+  final groupBy =
+      GroupAssetsBy.values[settings.getSetting(AppSettingsEnum.groupAssetsBy)];
+  yield await RenderList.fromQuery(query, groupBy);
+  await for (final _ in query.watchLazy()) {
+    yield await RenderList.fromQuery(query, groupBy);
+  }
```

Comment:
I saw this implementation across the PR. Am I understanding correctly that on line 175 we return the result the first time this provider is watched/called? Then we setup the a listener to the subsequent changes on line 176-178 for any future changes?

Reply:
Yeah, that is correct!